src/llm/io_processing/gptoss/reasoning_parser.cpp

```diff
@@ -0,0 +1,110 @@
+//*****************************************************************************
+// Copyright 2025 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include <openvino/genai/tokenizer.hpp>
+#include <string>
+#include <vector>
+
+#pragma warning(push)
+#pragma warning(disable : 6313)
+#include <rapidjson/document.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/writer.h>
+#pragma warning(pop)
+
+#include "../../../logging.hpp"
+#include "../../../stringutils.hpp"
+#include "reasoning_parser.hpp"
+#include "harmony.hpp"
+#include "../utils.hpp"
+
+#include "harmony.hpp"
+namespace ovms {
+void GptReasoningParser::parse(ParsedOutput& parsedOutput, const std::vector<int64_t>& generatedTokens) {
+    openai::Harmony harmony(tokenizer, generatedTokens);
+    if (!harmony.parse()) {
+        SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Harmony parsing failed");
+        return;
+    }
+
+    // TODO: How to enforce users to select reasoning parser even if they do not need reasoning?
+    parsedOutput.content = harmony.getContent();
+    SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Unary | GPT Content | [{}]", parsedOutput.content);
+    parsedOutput.reasoning = harmony.getReasoning();
+    SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Unary | GPT Reasoning | [{}]", parsedOutput.reasoning);
+}
+
+std::optional<rapidjson::Document> GptReasoningParser::parseChunk(const std::string& newChunk, ov::genai::GenerationFinishReason finishReason) {
+    SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Streaming | GPT Reason | Processing Chunk [{}]", newChunk);
+
+    if (newChunk.empty()) {
+        return std::nullopt;
+    }
+
+    std::string chunk = newChunk;
+
+    StreamState lastState = state;
+
+    if (startsWith(chunk, getParsingStartTag())) {
+        state = StreamState::READING_REASONING;
+        // remove the start tag
+        chunk = chunk.substr(getParsingStartTag().size());
+    } else if (startsWith(chunk, "<|start|>assistant<|channel|>final<|message|>")) {
```

Comment:
In special start tags I can see:
`"<|channel|>commentary<|message|>",  // Preable to reasoning, users usually sees that`
Doesn't it also require handling here?